lib/spack/spack/solver/asp.py

```diff
@@ -878,6 +878,31 @@ def package_compiler_defaults(self, pkg):
                 fn.node_compiler_preference(pkg.name, cspec.name, cspec.version, -i * 100)
             )
 
+    def cfg_required_rules(self, pkg):
+        pkg_name = pkg.name
+        config = spack.config.get("packages")
+        requirements = config.get(pkg_name, {}).get("require", {})
+        if isinstance(requirements, string_types):
+            rules = [(pkg_name, [requirements])]
+        else:
+            rules = []
+            for requirement in requirements:
+                rules.append((pkg_name, requirement["one_of"]))
+        cnd_grp_id = 0
+        for pkg_name, cnd_group in rules:
+            requirement_weight = 0
+            self.gen.fact(fn.requirement_group(pkg_name, cnd_grp_id))
+            for spec_str in cnd_group:
+                spec = spack.spec.Spec(spec_str)
+                if not spec.name:
+                    spec.name = pkg_name
```

Comment:
This is interesting, as it means we can do things like:
```yaml
packages:
  hdf5:
    require:
    - one_of: ['trilinos@X.Y']
```
where `trilinos` can be an arbitrary package. It seems fine to me, wondering if there are "smart" ways to use this.